File: frontend/src/config/http-gateway/room-service.js
import { AxiosClient } from './http-client';

export const RoomService = {
    /**
     * Obtiene todas las habitaciones
     * @returns {Promise} Lista de habitaciones
     */
    getAllRooms: async () => {
        try {
            const response = await AxiosClient.get('api/rooms');
            return { data: response, error: null };
        } catch (error) {
            return { data: null, error: error.response?.data || error.message };
        }
    },

    /**
     * Obtiene una habitación por ID
     * @param {number} id - ID de la habitación
     * @returns {Promise} Datos de la habitación
     */
    getRoomById: async (id) => {
        try {
            const response = await AxiosClient.get(`api/rooms/${id}`);
            return { data: response, error: null };
        } catch (error) {
            return { data: null, error: error.response?.data || error.message };
        }
    },

    /**
     * Obtiene habitaciones por estado
     * @param {string} status - Estado (EN_USO, PENDIENTE_LIMPIEZA, LIMPIA, BLOQUEADA_INCIDENCIA)
     * @returns {Promise} Lista de habitaciones con ese estado
     */
    getRoomsByStatus: async (status) => {
        try {
            const response = await AxiosClient.get(`api/rooms/status/${status}`);
            return { data: response, error: null };
        } catch (error) {
            return { data: null, error: error.response?.data || error.message };
        }
    },

    /**
     * Obtiene habitaciones por piso (Admin)
     * @param {string} floor - Piso
     * @returns {Promise} Lista de habitaciones del piso
     */
    getRoomsByFloor: async (floor) => {
        try {
            const response = await AxiosClient.get(`api/rooms/floor/${floor}`);
            return { data: response, error: null };
        } catch (error) {
            return { data: null, error: error.response?.data || error.message };
        }
    },

    /**
     * Crea una nueva habitación (Admin)
     * @param {Object} roomData - { roomNumber, floor?, barcodeValue?, notes? }
     * @returns {Promise} Habitación creada
     */
    createRoom: async (roomData) => {
        try {
            const response = await AxiosClient.post('api/rooms', roomData);
            return { data: response, error: null };
        } catch (error) {
            return { data: null, error: error.response?.data || error.message };
        }
    },

    /**
     * Actualiza una habitación (Admin)
     * @param {number} id - ID de la habitación
     * @param {Object} roomData - { roomNumber, floor?, barcodeValue?, notes? }
     * @returns {Promise} Habitación actualizada
     */
    updateRoom: async (id, roomData) => {
        try {
            const response = await AxiosClient.put(`api/rooms/${id}`, roomData);
            return { data: response, error: null };
        } catch (error) {
            return { data: null, error: error.response?.data || error.message };
        }
    },

    /**
     * Elimina una habitación (Admin)
     * @param {number} id - ID de la habitación
     * @returns {Promise}
     */
    deleteRoom: async (id) => {
        try {
            const response = await AxiosClient.delete(`api/rooms/${id}`);
            return { data: response, error: null };
        } catch (error) {
            return { data: null, error: error.response?.data || error.message };
        }
    },

    /**
     * Actualiza el estado de una habitación (Admin)
     * @param {number} id - ID de la habitación
     * @param {Object} statusData - { status, notes? }
     * @returns {Promise} Habitación con estado actualizado
     */
    updateRoomStatus: async (id, statusData) => {
        try {
            const response = await AxiosClient.patch(`api/rooms/${id}/status`, statusData);
            return { data: response, error: null };
        } catch (error) {
            return { data: null, error: error.response?.data || error.message };
        }
    }
};

// Constantes de estados de habitación
export const RoomStatus = {
    EN_USO: 'EN_USO',
    PENDIENTE_LIMPIEZA: 'PENDIENTE_LIMPIEZA',
    EN_LIMPIEZA: 'EN_LIMPIEZA',
    LIMPIA: 'LIMPIA',
    BLOQUEADA_INCIDENCIA: 'BLOQUEADA_INCIDENCIA'
};

File: backend/src/main/java/utez/edu/mx/backend/hotel/service/RoomService.java
package utez.edu.mx.backend.hotel.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import utez.edu.mx.backend.hotel.dto.RoomRequest;
import utez.edu.mx.backend.hotel.dto.RoomResponse;
import utez.edu.mx.backend.hotel.dto.UpdateRoomStatusRequest;
import utez.edu.mx.backend.hotel.model.HotelSettings;
import utez.edu.mx.backend.hotel.model.Room;
import utez.edu.mx.backend.hotel.repository.HotelSettingsRepository;
import utez.edu.mx.backend.hotel.repository.RoomAssignmentRepository;
import utez.edu.mx.backend.hotel.repository.RoomRepository;

import java.time.Instant;
import java.time.LocalTime;
import java.util.List;

@Service
@RequiredArgsConstructor
@Slf4j
public class RoomService {
    private final RoomRepository roomRepository;
    private final RoomAssignmentRepository roomAssignmentRepository;
    private final HotelSettingsRepository hotelSettingsRepository;

    @Transactional(readOnly = true)
    public List<RoomResponse> getAllRooms() {
        log.info("Obteniendo todas las habitaciones.");
        return roomRepository.findAll()
                .stream()
                .map(RoomResponse::fromEntity)
                .toList();
    }

    @Transactional(readOnly = true)
    public RoomResponse getRoomById(Long id) {
        log.info("Obteniendo habitación por ID: {}", id);
        Room room = roomRepository.findById(id)
                .orElseThrow(() -> {
                    log.error("Habitación no encontrada con ID: {}", id);
                    return new IllegalArgumentException("Habitación no encontrada");
                });
        return RoomResponse.fromEntity(room);
    }

    @Transactional
    public RoomResponse createRoom(RoomRequest request) {
        log.info("Iniciando creación de nueva habitación con número: {}", request.roomNumber());
        LocalTime defaultCleaningTime = hotelSettingsRepository.findAll()
                .stream()
                .findFirst()
                .map(HotelSettings::getCleaningStartTime)
                .orElse(LocalTime.of(14, 0));

        LocalTime cleaningTime = request.scheduledCleaningTime() != null
                ? request.scheduledCleaningTime()
                : defaultCleaningTime;

        Room room = Room.builder()
                .roomNumber(request.roomNumber())
                .floor(request.floor())
                .barcodeValue(request.barcodeValue())
                .scheduledCleaningTime(cleaningTime)
                .notes(request.notes())
                .currentStatus(Room.RoomStatus.LIMPIA)
                .lastStatusChange(Instant.now())
                .build();

        Room savedRoom = roomRepository.save(room);
        log.info("Habitación creada exitosamente con ID: {}", savedRoom.getId());
        return RoomResponse.fromEntity(savedRoom);
    }

    @Transactional
    public RoomResponse updateRoom(Long id, RoomRequest request) {
        log.info("Iniciando actualización de la habitación ID: {}", id);
        Room room = roomRepository.findById(id)
                .orElseThrow(() -> {
                    log.error("Error al actualizar: Habitación no encontrada con ID: {}", id);
                    return new IllegalArgumentException("Habitación no encontrada");
                });

        room.setRoomNumber(request.roomNumber());
        room.setFloor(request.floor());
        room.setBarcodeValue(request.barcodeValue());
        room.setScheduledCleaningTime(request.scheduledCleaningTime());
        room.setNotes(request.notes());

        Room updatedRoom = roomRepository.save(room);
        log.info("Habitación ID: {} actualizada exitosamente.", id);
        return RoomResponse.fromEntity(updatedRoom);
    }

    @Transactional
    public void deleteRoom(Long id) {
        log.warn("Iniciando eliminación de la habitación ID: {}", id);
        if (!roomRepository.existsById(id)) {
            log.error("Error al eliminar: Habitación no encontrada con ID: {}", id);
            throw new IllegalArgumentException("Habitación no encontrada");
        }
        roomRepository.deleteById(id);
        log.info("Habitación ID: {} eliminada exitosamente.", id);
    }

    @Transactional
    public RoomResponse updateRoomStatus(Long id, UpdateRoomStatusRequest request) {
        log.info("Iniciando actualización de estado para la habitación ID: {} a {}", id, request.status());
        Room room = roomRepository.findById(id)
                .orElseThrow(() -> {
                    log.error("Error al actualizar estado: Habitación no encontrada con ID: {}", id);
                    return new IllegalArgumentException("Habitación no encontrada");
                });

        room.setCurrentStatus(request.status());
        room.setLastStatusChange(Instant.now());

        Room updatedRoom = roomRepository.save(room);
        log.info("Estado de la habitación ID: {} actualizado a {}.", id, request.status());

        // Si la habitación se marca como LIMPIA, desactivar automáticamente sus asignaciones activas
        if (request.status() == Room.RoomStatus.LIMPIA) {
            log.info("La habitación ID: {} se marcó como LIMPIA. Desactivando asignaciones activas.", id);
            roomAssignmentRepository.findByRoomIdAndActiveTrue(id)
                    .forEach(assignment -> {
                        assignment.setActive(false);
                        roomAssignmentRepository.save(assignment);
                        log.debug("Asignación ID: {} desactivada para la habitación ID: {}", assignment.getId(), id);
                    });
        }

        return RoomResponse.fromEntity(updatedRoom);
    }

    @Transactional(readOnly = true)
    public List<RoomResponse> getRoomsByStatus(Room.RoomStatus status) {
        log.info("Obteniendo habitaciones con estado: {}", status);
        return roomRepository.findByCurrentStatus(status)
                .stream()
                .map(RoomResponse::fromEntity)
                .toList();
    }

    @Transactional(readOnly = true)
    public List<RoomResponse> getRoomsByFloor(String floor) {
        log.info("Obteniendo habitaciones del piso: {}", floor);
        return roomRepository.findByFloor(floor)
                .stream()
                .map(RoomResponse::fromEntity)
                .toList();
    }
}
